### PR TITLE
KEYCLOAK-6749 Jetty App Server

### PR DESCRIPTION
**Pull request by slaskawi**
https://issues.jboss.org/browse/KEYCLOAK-6749

This PR introduces Jetty App Server support. 

Still to be done:

* [x] Implement pipeline pull request
* [x] Remove old tests

### CODE REVIEW

File: testsuite/integration-arquillian/servers/app-server/jetty/common/pom.xml
Comment (slaskawi) [inline testsuite/integration-arquillian/servers/app-server/jetty/common/pom.xml]: Here's the interesting part for you guys.

As you can see the `common` module consists of some mandatory dependencies (like the `keycloak-saml-servlet-filter-adapter`) and a set of dependencies marked as `<optional>true</optional>`. Optional dependencies are not transitive. Therefore, it is up to `integration-arquillian-servers-app-server-jetty-9x` (and `integration-arquillian-servers-app-server-jetty-81`) modules to set the dependencies properly but the implementation is shared at the same time. If you want to see this in action, please have a look that I'm using a different set of dependencies for 8 and 9 around security as well as JSPs. 

Similar thing happens to the tests. As you can see, the `common` module contains basic "sanity" tests (just to check if app server implementation looks OK-ish), which are reused by specific modules.

File: testsuite/integration-arquillian/servers/app-server/jetty/81/pom.xml
Comment (vramik) [inline testsuite/integration-arquillian/servers/app-server/jetty/81/pom.xml]: nit: the explicit version declaration could be removed here as the managed version is inherited from parent pom

File: testsuite/integration-arquillian/servers/app-server/jetty/81/src/test/java/org/keycloak/testsuite/arquillian/jetty/Jetty81AppServerTest.java
Comment (vramik) [inline testsuite/integration-arquillian/servers/app-server/jetty/81/src/test/java/org/keycloak/testsuite/arquillian/jetty/Jetty81AppServerTest.java]: I get 

```
Tests in error: 
  Jetty81AppServerTest>AbstractJettyAppServerTest.testDeployingAndUndeploying:123 » NoClassDefFound
  Jetty81AppServerTest>AbstractJettyAppServerTest.testDeployingRESTApp:95 » NoClassDefFound
  Jetty81AppServerTest>AbstractJettyAppServerTest.testDeployingServletApp:67 » NoClassDefFound
  Jetty81AppServerTest>AbstractJettyAppServerTest.testServerStartupAndShutdown:43 » NoClassDefFound

...
java.lang.NoClassDefFoundError: org/eclipse/jetty/server/HttpConfiguration$Customizer
...
```

when I'm trying to build `integration-arquillian-servers-app-server`

File: testsuite/integration-arquillian/servers/app-server/jetty/common/pom.xml
Comment (vramik) [inline testsuite/integration-arquillian/servers/app-server/jetty/common/pom.xml]: explicit version declaration

File: testsuite/integration-arquillian/servers/app-server/jetty/common/pom.xml
Comment (vramik) [inline testsuite/integration-arquillian/servers/app-server/jetty/common/pom.xml]: explicit version declaration

File: testsuite/integration-arquillian/servers/app-server/jetty/pom.xml
Comment (vramik) [inline testsuite/integration-arquillian/servers/app-server/jetty/pom.xml]: I think we should include the modules in profile definition, it would build all jetty app servers each time the testsuite is built. 

Check https://github.com/keycloak/keycloak/blob/22e160136c37d73b05b7f4d77de5a2d1fcb7cef5/testsuite/integration-arquillian/servers/app-server/jboss/pom.xml#L710-L785

File: testsuite/integration-arquillian/servers/app-server/undertow/pom.xml
Comment (vramik) [inline testsuite/integration-arquillian/servers/app-server/undertow/pom.xml]: Isn't this a mistake?

File: testsuite/integration-arquillian/tests/base/pom.xml
Comment (vramik) [inline testsuite/integration-arquillian/tests/base/pom.xml]: Do we plan to test all jetty versions in one test run? If not there should me imho jetty81, shouldn't be?

File: testsuite/integration-arquillian/tests/base/pom.xml
Comment (vramik) [inline testsuite/integration-arquillian/tests/base/pom.xml]: the same as above

**Review by vramik — CHANGES_REQUESTED** [review]
Thank you @slaskawi for the PR. I have few question or comments.

File: testsuite/integration-arquillian/servers/app-server/jetty/81/src/test/java/org/keycloak/testsuite/arquillian/jetty/Jetty81AppServerTest.java
Comment (slaskawi) [reply testsuite/integration-arquillian/servers/app-server/jetty/81/src/test/java/org/keycloak/testsuite/arquillian/jetty/Jetty81AppServerTest.java]: Yes, I accidentally broke it just before the push. 

Yesterday I spoke with @hmlnarik that we won't migrate 8.1 version to the new testsuite (due to [KEYCLOAK-8356](https://issues.jboss.org/browse/KEYCLOAK-8356)). I will remove this part shortly.

File: testsuite/integration-arquillian/tests/base/pom.xml
Comment (slaskawi) [reply testsuite/integration-arquillian/tests/base/pom.xml]: This is very good question...

All Jetty implementation share the same code. What changes between them, are dependencies. So far I've been using `@AppServerContainer(ContainerConstants.APP_SERVER_JETTY)`, which means the testing code doesn't differentiate Jetty versions. So far this works fine, so I don't see any reason to introduce specific versions in Jetty App Server. I hope this won't change between Jetty 9.4, 9.3 and 9.2 implementations (as 9.3 and 9.2 are not there yet). 

So unless there's any good reason, perhaps we could stay with `app.server=jetty` without any version? 

@mhajas @hmlnarik @vramik WDYT?

File: testsuite/integration-arquillian/tests/base/pom.xml
Comment (slaskawi) [reply testsuite/integration-arquillian/tests/base/pom.xml]: Let's continue the discussion a bit higher in this thread. See https://github.com/keycloak/keycloak/pull/5885#discussion_r257982873

File: testsuite/integration-arquillian/servers/app-server/jetty/81/pom.xml
Comment (slaskawi) [reply testsuite/integration-arquillian/servers/app-server/jetty/81/pom.xml]: Fixed. Thanks!

File: testsuite/integration-arquillian/servers/app-server/jetty/common/pom.xml
Comment (slaskawi) [reply testsuite/integration-arquillian/servers/app-server/jetty/common/pom.xml]: Fixed. Thanks!

File: testsuite/integration-arquillian/servers/app-server/jetty/common/pom.xml
Comment (slaskawi) [reply testsuite/integration-arquillian/servers/app-server/jetty/common/pom.xml]: Fixed. Thanks!

File: testsuite/integration-arquillian/servers/app-server/undertow/pom.xml
Comment (slaskawi) [reply testsuite/integration-arquillian/servers/app-server/undertow/pom.xml]: Fixed. Thanks!

File: testsuite/integration-arquillian/tests/base/pom.xml
Comment (vramik) [reply testsuite/integration-arquillian/tests/base/pom.xml]: @slaskawi thanks. 

If you stay with `app.server=jetty` without any version, how do you know which version of jetty server should be started? Or do you plan to test all version in one test run? In that case you'd need to rerun every single test for every jetty version. Something like `@ParameterizedTest` introduced in JUnit 5? Maybe I'm missing something. 

I'll do the review again with you latest changes and see if I can understand it better. Is it ready for the review now?

File: testsuite/integration-arquillian/servers/app-server/jetty/pom.xml
Comment (hmlnarik) [reply testsuite/integration-arquillian/servers/app-server/jetty/pom.xml]: +1

File: testsuite/integration-arquillian/tests/base/pom.xml
Comment (hmlnarik) [reply testsuite/integration-arquillian/tests/base/pom.xml]: APP_SERVER_ annotations should correspond to available app server modules, i.e. should differentiate between the version. APP_SERVER_JETTY81, APP_SERVER_JETTY92, APP_SERVER_JETTY93, APP_SERVER_JETTY94. Cf. [APP_SERVER_WILDFLY](https://github.com/keycloak/keycloak/blob/24b7d080af6ea554e346aa4ab770f00450696e4d/testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/arquillian/containers/ContainerConstants.java#L24) and [APP_SERVER_WILDFLY_DEPRECATED](https://github.com/keycloak/keycloak/blob/24b7d080af6ea554e346aa4ab770f00450696e4d/testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/arquillian/containers/ContainerConstants.java#L25):

File: testsuite/integration-arquillian/servers/app-server/jetty/pom.xml
Comment (slaskawi) [reply testsuite/integration-arquillian/servers/app-server/jetty/pom.xml]: I'm not sure if this isn't a bit over the top for Jetty. All modules build in less than a second:

```
[INFO] App Server - Jetty Parent .......................... SUCCESS [  0.024 s]
[INFO] App Server - Jetty Common .......................... SUCCESS [  0.512 s]
[INFO] App Server - Jetty 9.2 ............................. SUCCESS [  0.196 s]
[INFO] App Server - Jetty 9.3 ............................. SUCCESS [  0.195 s]
[INFO] App Server - Jetty 9.4 ............................. SUCCESS [  0.161 s]
```

Please let me know if you still would like to have this.

File: testsuite/integration-arquillian/tests/base/pom.xml
Comment (slaskawi) [reply testsuite/integration-arquillian/tests/base/pom.xml]: Ok, great! Thanks for the explanation. I'll try to introduce necessary changes shortly.

File: testsuite/integration-arquillian/servers/app-server/jetty/pom.xml
Comment (vramik) [reply testsuite/integration-arquillian/servers/app-server/jetty/pom.xml]: But have you tested it without the dependencies downloaded to your local maven repo? I know the server is not that big as a wildfly but still.

File: testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/arquillian/AdapterTestExecutionDecider.java
Comment (slaskawi) [inline testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/arquillian/AdapterTestExecutionDecider.java]: This change is fun... hold on tight!

The cache in this method doesn't work properly in a situation we have in this PR, such as `DemoServletsAdapterTest` runs on several app servers, whereas `JettyDemoServletsAdapterTest` runs on a single server which is not mentioned in `DemoServletsAdapterTest`. If `DemoServletsAdapterTest` is executed first, it caches, which methods should run. Let's assume we test with Undertow app server, so the result is `true` for all methods. Next, we execute `JettyDemoServletsAdapterTest` again with Undertow app server. Since, this class is for Jetty, the testsuite shouldn't execute any of these methods (remember, we assumed app server is Undertow). However it tries to do that...

Let me explain why...

The `Method#equals` method uses declaring class (and parameters) to do equality check. Note that the declaring (!!!) method is still the same for both classes - it's `DemoServletsAdapterTest`. This is why it reuses cached entry from `DemoServletsAdapterTest` run instead of re-evaluating whether the test should run or not. Also note, that this caching may introduce more damage that we are not aware of. There are some methods that possible were not executed in our testsuite (or were executed but shouldn't, but if they passed, that's fine). It all depends on the order that Surefire picks.

I decided to remove the cache altogether. What we do in this method is checking a bunch of `if` statements plus traversing test class hierarchy. Both things are blazing fast on modern JVM and CPUs.

File: testsuite/integration-arquillian/servers/app-server/jetty/pom.xml
Comment (slaskawi) [reply testsuite/integration-arquillian/servers/app-server/jetty/pom.xml]: This is a very good point actually. Yes, I'll implement this shortly.

File: testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/arquillian/AdapterTestExecutionDecider.java
Comment (vramik) [reply testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/arquillian/AdapterTestExecutionDecider.java]: Maybe I don't get it right but afaik we test single app at the time, so in case we use app-server-undertow any other app server is not tested. The decider should decide if the method is testable or not for this particular container. 

I have nothing against removing the cache completely (as it has minimal impact like you say) in case the decider works as it supposed to.

File: testsuite/integration-arquillian/servers/app-server/jetty/pom.xml
Comment (slaskawi) [reply testsuite/integration-arquillian/servers/app-server/jetty/pom.xml]: Implemented. Thanks!

File: testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/arquillian/AdapterTestExecutionDecider.java
Comment (slaskawi) [reply testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/arquillian/AdapterTestExecutionDecider.java]: Yes, that's the case. The tricky part is that by default we executed with app server Undertow. However, `JettyDemoServletsAdapterTest` doesn't support it. That's why this tests shouldn't be executed but Surefire was trying to run it. I believe we didn't have similar situation before (Undertow is supported by all the adapters, Jetty is a bit special one).

Yes, I also checked the logs. They look fine. Thanks a lot for the comment!

File: testsuite/integration-arquillian/tests/base/pom.xml
Comment (slaskawi) [reply testsuite/integration-arquillian/tests/base/pom.xml]: Implemented thanks!

**Comment by slaskawi** [comment]
@keycloak-ci-bot test

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/jaas/LoginModulesTest.java
Comment (mhajas) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/jaas/LoginModulesTest.java]: Don't know those tests very well, but it might be the same case as JavascriptAdapterTest

File: testsuite/integration-arquillian/util/src/main/java/org/keycloak/testsuite/utils/tls/TLSUtils.java
Comment (mhajas) [inline testsuite/integration-arquillian/util/src/main/java/org/keycloak/testsuite/utils/tls/TLSUtils.java]: \* in imports again

**Review by mhajas — CHANGES_REQUESTED** [review]
Great work @slaskawi. Sorry for late review but I found few things I am not sure about. Could you please look at it?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/javascript/JavascriptAdapterTest.java
Comment (slaskawi) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/javascript/JavascriptAdapterTest.java]: This change has been specifically introduced for my Pull Request against the pipeline. Please have a quick glance at the before reading below.

Before this change, the Javascript adapter tests were invoked with each App Server. This was absolutely fine, since we tested only Wildfly App Server in the pipeline. However, with matrix-like approach to the adapter tests (we execute the adapter tests with different App Servers such as Wildfly, Jetty 9.2, Jetty 9.3 etc) this becomes wasteful (especially in terms of testing time). 

While working on this, I wasn't sure with what App Servers combinations we want to run Javascript adapter tests. I only knew, they don't make sense (and probably nobody expects them to run) with Jetty. That's why there are all App Servers on the list except the new one - Jetty. 

With your explanation, I think we can enhance this solution a bit more. Perhaps Javascript Adapter test should only run with the default App Server (Undertow). However, we would also need to add Undertow to the pipeline. If you agree with me, I would like to create a JIRA for it. We are approaching to the end of the sprint and my Pull Request against the pipeline hasn't been reviewed so far, so it's a bit risky to do this change now.

Alternatively, if you don't like this approach, I can remove all the annotations and allow this test to execute always. But as I mentioned before, it seems a bit wasteful to me.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/javascript/JavascriptAdapterWithNativePromisesTest.java
Comment (slaskawi) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/javascript/JavascriptAdapterWithNativePromisesTest.java]: Let's continue the conversation in https://github.com/keycloak/keycloak/pull/5885#discussion_r261115437. Once we agree on the approach, I'll apply it here as well.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/jaas/LoginModulesTest.java
Comment (slaskawi) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/jaas/LoginModulesTest.java]: Let's continue the conversation in https://github.com/keycloak/keycloak/pull/5885#discussion_r261115437. Once we agree on the approach, I'll apply it here as well.

File: adapters/oidc/jetty/jetty-core/src/main/java/org/keycloak/adapters/jetty/core/JettyCookieTokenStore.java
Comment (slaskawi) [reply adapters/oidc/jetty/jetty-core/src/main/java/org/keycloak/adapters/jetty/core/JettyCookieTokenStore.java]: Good point! Let me fix that.

File: testsuite/integration-arquillian/util/src/main/java/org/keycloak/testsuite/utils/tls/TLSUtils.java
Comment (slaskawi) [reply testsuite/integration-arquillian/util/src/main/java/org/keycloak/testsuite/utils/tls/TLSUtils.java]: Good point! Let me fix that.

File: adapters/oidc/jetty/jetty-core/pom.xml
Comment (slaskawi) [reply adapters/oidc/jetty/jetty-core/pom.xml]: Good point! Fixing.

**Comment by slaskawi** [comment]
This PR should be run with a dedicated update of the Pipeline (See the Pull Requests in our pipeline repository).

I created a customized job, to show you how it works (and to verify that I didn't break anything of course). I highly recommend merging those two things together.

See the Jenkins job here: https://keycloak-jenkins.com/blue/organizations/jenkins/slaskawi-keycloak-pr-pipeline-structured-jetty-adapters/detail/slaskawi-keycloak-pr-pipeline-structured-jetty-adapters/12/pipeline

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/javascript/JavascriptAdapterTest.java
Comment (mhajas) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/javascript/JavascriptAdapterTest.java]: I see you are right. Probably we should run it only with undertow, however, it might be a little bit confusing for others looking at it and say "why it only runs with undertow, do they even test Javascript adapter with wildfly?". 

Maybe we could create a separate package for adapter tests which can run without application server (maybe simply javascript package). This way we won't have the problem of multiple runs as it will be part of base testsuite. @slaskawi @vramik WDYT?

**Comment by vramik** [comment]
@keycloak-ci-bot test

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/javascript/JavascriptAdapterTest.java
Comment (slaskawi) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/javascript/JavascriptAdapterTest.java]: Probably the best option would be to create a new, special group for Javascript (and possibly other, non-app-server tests). I created [KEYCLOAK-9712](https://issues.jboss.org/browse/KEYCLOAK-9712) to implement this. 

Since it's getting late in this sprint, let me revert this change to Javascript tests (and let them always execute). We can revisit this thread once working on KEYCLOAK-9712.

Comment (mhajas) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/javascript/JavascriptAdapterTest.java]: ok I agree

**Comment by slaskawi** [comment]
Here's the new build for this pipeline: 
https://keycloak-jenkins.com/blue/organizations/jenkins/slaskawi-keycloak-pr-pipeline-structured-jetty-adapters/detail/slaskawi-keycloak-pr-pipeline-structured-jetty-adapters/13/pipeline

**Comment by slaskawi** [comment]
@keycloak-ci-bot test

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/jaas/LoginModulesTest.java
Comment (vramik) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/jaas/LoginModulesTest.java]: As agreed in https://github.com/keycloak/keycloak/pull/5885#discussion_r261115437 this should be also reverted (and maybe another JIRA created)

**Review by vramik — COMMENTED** [review]
Thank you @slaskawi, nice work! I have just two last small comments. Once those will be incorporated and commits squashed into one, it's good to go in IMHO.

**Comment by vramik** [comment]
https://github.com/keycloak/keycloak/blob/8dfd07662c114680dfdfa44254ebe9fddae83586/travis-run-tests.sh#L55 needs to be also deleted.

**Review by mhajas — DISMISSED** [review]
Great work @slaskawi  I like it. Just one minor comment

File: testsuite/integration-arquillian/pom.xml
Comment (slaskawi) [reply testsuite/integration-arquillian/pom.xml]: Fixed, thanks!

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/jaas/LoginModulesTest.java
Comment (slaskawi) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/jaas/LoginModulesTest.java]: Fixed. Thanks!

**Comment by slaskawi** [comment]
@vramik Regarding to `mvn test -B -nsu -f jetty`, I believe you looked at some outdated diff. This has already been removed.

File: testsuite/integration-arquillian/servers/app-server/jetty/92/pom.xml
Comment (slaskawi) [reply testsuite/integration-arquillian/servers/app-server/jetty/92/pom.xml]: Fixed. Thanks!

File: testsuite/integration-arquillian/servers/app-server/jetty/93/pom.xml
Comment (slaskawi) [reply testsuite/integration-arquillian/servers/app-server/jetty/93/pom.xml]: Fixed. Thanks!

File: testsuite/integration-arquillian/servers/app-server/jetty/94/pom.xml
Comment (slaskawi) [reply testsuite/integration-arquillian/servers/app-server/jetty/94/pom.xml]: Fixed. Thanks!

**Comment by vramik** [comment]
> @vramik Regarding to `mvn test -B -nsu -f jetty`, I believe you looked at some outdated diff. This has already been removed.

yeah, you're right, thanks

**Comment by slaskawi** [comment]
@keycloak-ci-bot test

**Review by vramik — DISMISSED** [review]
@slaskawi thank you for incorporating all my comments

**Comment by slaskawi** [comment]
Squashed, rebased and documentation added.

**Comment by slaskawi** [comment]
@keycloak-ci-bot test

**Review by hmlnarik — APPROVED** [review]

**Review by vramik — APPROVED** [review]

**Review by mhajas — APPROVED** [review]

File: testsuite/integration-arquillian/HOW-TO-RUN.md
Comment (mhajas) [inline testsuite/integration-arquillian/HOW-TO-RUN.md]: @slaskawi This is probably wrong. I think you need to either add another step in which you build integration-arquillian/servers/app-server/jetty or change this path to:

````
mvn -f testsuite/integration-arquillian/pom.xml
````

Comment (hmlnarik) [reply testsuite/integration-arquillian/HOW-TO-RUN.md]: You're right. Yet I believe this minor issue can be fixed later

Comment (slaskawi) [reply testsuite/integration-arquillian/HOW-TO-RUN.md]: Yes, it also requires a server module to be built. Probably I could change the path to `testsuite/integration-arquillian`. @mhajas Would that work for you? And would you guys like me to add it now? @mhajas @hmlnarik

Comment (mhajas) [reply testsuite/integration-arquillian/HOW-TO-RUN.md]: With @hmlnarik we decided not to do it in this PR. It is just a minor thing which can be fixed in my PR (Tomcat) so that we won't need to wait for another CI runs for this PR.

Comment (slaskawi) [reply testsuite/integration-arquillian/HOW-TO-RUN.md]: Ok, great! Thanks guys!

**Pull request merged**
